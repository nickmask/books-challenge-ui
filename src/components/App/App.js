import React from "react";
import { BrowserRouter as Router, Route, Switch, Link } from "react-router-dom";
import { FaArrowLeft, FaShoppingCart } from "react-icons/fa";

import BookList from "../BookList/BookList";
import BookDetails from "../BookDetails/BookDetails";
import Cart from "../Cart/Cart";

import styles from "./App.module.css";

class App extends React.Component {
  state = {
    cartItems: []
  };

  handleAddToCart = bookDetails => {
    this.setState(prevState => {
      const cartItems = [...prevState.cartItems, bookDetails];

      return { cartItems };
    });
  };

  getTotalCartValue = () => {
    const totalValue = this.state.cartItems.reduce(
      (accumulator, currentBook) => currentBook.price + accumulator,
      0
    );

    return totalValue;
  };

  removeCartItem = index => {
    this.setState(prevState => {
      const cartItems = [...prevState.cartItems];
      cartItems.splice(index, 1);

      return { cartItems };
    });
  };

  render() {
    const { cartItems } = this.state;
    const backToBooks = (
      <Link className={styles.back} to="/">
        <FaArrowLeft />
        <span className={styles.backText}>back to books</span>
      </Link>
    );

    return (
      <Router>
        <div className={styles.container}>
          <header className={styles.header}>
            <span>Books on tap</span>
          </header>
          <nav className={styles.nav}>
            <Switch>
              <Route exact path="/" render={() => <span />} />
              <Route path="/book" render={() => backToBooks} />
              <Route path="/cart" render={() => backToBooks} />
            </Switch>

            <Link to="/cart">
              <div className={styles.cart}>
                <p aria-label="Total cart value">{`Â£${this.getTotalCartValue()}`}</p>
                <hr className={styles.seperator} />
                <p
                  aria-label="Number of items in cart"
                  className={styles.cartCount}
                >
                  <FaShoppingCart className={styles.cartIcon} color={"#fff"} />
                  <span>{cartItems.length}</span>
                </p>
              </div>
            </Link>
          </nav>

          <main>
            <Switch>
              <Route exact path="/" component={BookList} />
              <Route
                exact
                path="/book/:title/:id"
                render={routeProps => (
                  <BookDetails
                    {...routeProps}
                    onAddToCart={this.handleAddToCart}
                    cartItems={cartItems}
                  />
                )}
              />
              <Route
                exact
                path="/cart"
                render={routeProps => (
                  <Cart
                    {...routeProps}
                    totalValue={this.getTotalCartValue()}
                    cartItems={cartItems}
                    onRemoveCartItem={this.removeCartItem}
                  />
                )}
              />
              <Route render={() => <h1>404</h1>} />
            </Switch>
          </main>
          <footer />
        </div>
      </Router>
    );
  }
}

export default App;
