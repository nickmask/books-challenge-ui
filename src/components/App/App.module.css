.container {
  display: grid;

  grid-template-areas:
    "header header header header header"
    "nav nav nav nav nav"
    ". leftSidebar content rightSidebar ."
    "footer footer footer footer footer";

  grid-template-columns: auto auto minmax(auto, 980px) auto auto;
  grid-template-rows: auto auto 1fr auto;

  height: 100vh;
}

@media (max-width: 768px) {
  .container {
    grid-template-areas:
      "header"
      "nav"
      "content"
      "footer";

    grid-template-columns: 1fr;
    grid-template-rows:
      auto /* Header */
      auto /* Nav */
      1fr /* Content */
      auto; /* Footer */
  }
}

.header {
  align-items: center;
  background-color: #273456;
  color: #fff;
  display: flex;
  font-size: 2rem;
  height: 75px;
  padding-left: 20px;
}

.nav {
  display: flex;
  justify-content: space-between;
  align-items: center;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.2);
  padding: 0 20px;
}

.cart {
  align-items: center;
  background-color: #f2c14a;
  display: flex;
  flex-direction: row;
  height: 40px;
  justify-content: center;
  width: 140px;
  color: #fff;
}

.cart > * {
  margin: 0 10px;
}

.cartCount {
  display: flex;
  justify-content: center;
}

.cartIcon {
  margin-right: 5px;
}

.seperator {
  border: none;
  border-left: 1px solid rgba(255, 255, 255, 0.4);
  height: 22px;
  width: 1px;
  margin: 0 10px;
}

.back {
  display: flex;
  align-items: center;
}

.backText {
  padding-left: 10px;
}
