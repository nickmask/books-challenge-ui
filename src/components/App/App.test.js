import React from "react";
import { shallow } from "enzyme";
import createRouterContext from "react-router-test-context";

import App from "./App";

describe("App", () => {
  const context = createRouterContext();

  it("should render correctly", () => {
    const component = shallow(<App />, { context });

    expect(component).toMatchSnapshot();
  });

  it("should get correct total cart value", () => {
    const component = shallow(<App />, { context });

    component.setState({
      cartItems: [
        { id: 1, price: 20 },
        { id: 1, price: 15 },
        { id: 2, price: 20 },
        { id: 99, price: 50 }
      ]
    });

    const getTotalCartValueMethod = component.instance().getTotalCartValue;

    expect(getTotalCartValueMethod()).toEqual(105);
  });

  it("should get correct total cart when empty", () => {
    const component = shallow(<App />, { context });

    component.setState({
      cartItems: []
    });

    const getTotalCartValueMethod = component.instance().getTotalCartValue;

    expect(getTotalCartValueMethod()).toEqual(0);
  });

  it("should get correct total cart value with floats", () => {
    const component = shallow(<App />, { context });

    component.setState({
      cartItems: [
        { id: 1, price: 20 },
        { id: 1, price: 15.7 },
        { id: 2, price: 20 },
        { id: 99, price: 50.5 }
      ]
    });

    const getTotalCartValueMethod = component.instance().getTotalCartValue;

    expect(getTotalCartValueMethod()).toEqual(106.2);
  });

  it("should handle adding books to cart", () => {
    const component = shallow(<App />, { context });

    const handleAddToCart = component.instance().handleAddToCart;

    handleAddToCart({ id: 1, name: "The Great Booksy", price: 20 });

    expect(component.state().cartItems).toEqual([
      { id: 1, name: "The Great Booksy", price: 20 }
    ]);

    handleAddToCart({ id: 2, name: "Notting Shill", price: 20 });

    expect(component.state().cartItems).toEqual([
      { id: 1, name: "The Great Booksy", price: 20 },
      { id: 2, name: "Notting Shill", price: 20 }
    ]);
  });
});
