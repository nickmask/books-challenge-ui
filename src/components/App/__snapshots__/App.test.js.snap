// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<BrowserRouter>
  <div
    className="container"
  >
    <header
      className="header"
    >
      <span>
        Books on tap
      </span>
    </header>
    <nav
      className="nav"
    >
      <Switch>
        <Route
          exact={true}
          path="/"
          render={[Function]}
        />
        <Route
          path="/book"
          render={[Function]}
        />
        <Route
          path="/cart"
          render={[Function]}
        />
      </Switch>
      <Link
        to="/cart"
      >
        <div
          className="cart"
        >
          <p
            aria-label="Total cart value"
          >
            Â£0
          </p>
          <hr
            className="seperator"
          />
          <p
            aria-label="Number of items in cart"
            className="cartCount"
          >
            <FaShoppingCart
              className="cartIcon"
              color="#fff"
            />
            <span>
              0
            </span>
          </p>
        </div>
      </Link>
    </nav>
    <main>
      <Switch>
        <Route
          component={[Function]}
          exact={true}
          path="/"
        />
        <Route
          exact={true}
          path="/book/:title/:id"
          render={[Function]}
        />
        <Route
          exact={true}
          path="/cart"
          render={[Function]}
        />
        <Route
          render={[Function]}
        />
      </Switch>
    </main>
    <footer />
  </div>
</BrowserRouter>
`;
